Add coverage workflow

tequdev opened the pull request:
<!--
This PR template helps you to write a good pull request description.
Please feel free to include additional useful information even beyond what is requested below.
-->

## High Level Overview of Change

<!--
Please include a summary of the changes.
This may be a direct input to the release notes.
If too broad, please consider splitting into multiple PRs.
If a relevant task or issue, please link it here.
-->

### Context of Change

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a spec or design document for this feature, please link it here.
-->

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

<!--
Please check [x] relevant options, delete irrelevant ones.

* If there is any impact to the public API methods (HTTP / WebSocket), please update https://github.com/xrplf/rippled/blob/develop/API-CHANGELOG.md
  * Update API-CHANGELOG.md and add the change directly in this PR by pushing to your PR branch.
* libxrpl: See https://github.com/XRPLF/rippled/blob/develop/docs/build/depend.md
* Peer Protocol: See https://xrpl.org/peer-protocol.html
-->

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [ ] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

<!--
## Before / After
If relevant, use this section for an English description of the change at a technical level.
If this change affects an API, examples should be included here.
-->

<!--
## Test Plan
If helpful, please describe the tests that you ran to verify your changes and provide instructions so that others can reproduce.
This section may not be needed if your change includes thoroughly commented unit tests.
-->

<!--
## Future Tasks
For future tasks related to PR.
-->

Bronek committed:
test: improve code coverage reporting (#4849)

* Speed up the generation of coverage reports by using multiple cores.

* Add codecov step to coverage workflow.

tequdev committed:
Add coverage workflow

tequdev committed:
fix cmake-target

tequdev committed:
Merge branch 'dev' into coverage

tequdev committed:
Merge branch 'dev' into coverage

tequdev committed:
Merge branch 'dev' into coverage

tequdev committed:
Merge branch 'dev' into coverage

sublimator commented:
### Bug: `--target` ordering with `VERBOSE_FLAG`

In `.github/actions/xahau-ga-build/action.yml:228`:

```bash
cmake --build . --config ... --parallel $(nproc) ${VERBOSE_FLAG} --target ${{ inputs.cmake-target }}
```

When someone uses `[ci-ga-cmake-verbose]` in a commit message, `VERBOSE_FLAG` becomes `-- -v`, which means `--target coverage` gets passed to the native build tool (Ninja/Make) instead of CMake. Fix is to move `--target` before `${VERBOSE_FLAG}`:

```bash
cmake --build . --config ... --parallel $(nproc) --target ${{ inputs.cmake-target }} ${VERBOSE_FLAG}
```

sublimator committed:
fix: ci build action and coverage workflow fixes (#682)

- move --target before ${VERBOSE_FLAG} so [ci-ga-cmake-verbose] doesn't
  swallow the target argument
- add stdlib: default option for GCC entries (GCC doesn't support -stdlib flag)
- pass stdlib explicitly to coverage build step
- remove dead CMAKE_BUILD_PARALLEL_LEVEL export (doesn't persist across steps,
  and build step already uses --parallel)
- remove duplicate ccache config block

sublimator committed:
Merge branch 'dev' into coverage

RichardAH committed:
Merge branch 'dev' into coverage

tequdev committed:
Merge branch 'dev' into coverage

tequdev committed:
Merge branch 'dev' into coverage

tequdev committed:
Merge branch 'dev' into coverage

RichardAH committed:
Merge branch 'dev' into coverage
